File: playnow-frontend/src/app/core/services/booking.service.ts
import { Injectable, signal, computed } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { AuthService } from './auth.service';

export interface Booking {
  id: string;
  playerName: string;
  turfId: string;
  turfName: string;
  date: string;
  timeSlot: string;
  duration: number;
  teamSize: number;
  totalCost: number;
  status: 'pending' | 'confirmed' | 'cancelled';
  createdAt: Date;
}

export interface Turf {
  id: string;
  name: string;
  location: string;
  pricePerHour: number;
  isAvailable: boolean;
  imageUrl?: string;
}

// Backend DTOs
export interface CreateBookingRequest {
  turfId: number;
  bookingDate: string;
  startTime: string;
  endTime: string;
  specialRequests?: string;
}

export interface BookingResponse {
  id: number;
  turfId: number;
  turfName: string;
  bookingDate: string;
  startTime: string;
  endTime: string;
  durationHours: number;
  totalAmount: number;
  status: string;
  specialRequests?: string;
  createdAt: string;
  payment?: PaymentResponse;
}

export interface PaymentResponse {
  id: number;
  paymentMethod: string;
  transactionId: string;
  amount: number;
  status: string;
  paymentDate: string;
}

export interface TurfResponse {
  id: number;
  name: string;
  description: string;
  address: string;
  city: string;
  sportType: string;
  capacity: number;
  pricePerHour: number;
  imageUrl?: string;
  isAvailable: boolean;
  amenities: string[];
}

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  // Using Angular 17 Signals for reactive state management
  private bookings = signal<Booking[]>([]);
  private turfs = signal<Turf[]>([]);
  private isLoading = signal(false);

  private readonly API_BASE_URL = 'http://localhost:5048/api';

  // Computed signals for derived state
  public readonly allBookings = this.bookings.asReadonly();
  public readonly allTurfs = this.turfs.asReadonly();
  public readonly isLoadingState = this.isLoading.asReadonly();
  
  public readonly availableTurfs = computed(() => 
    this.turfs().filter(turf => turf.isAvailable)
  );

  public readonly totalBookings = computed(() => 
    this.bookings().length
  );

  public readonly confirmedBookings = computed(() => 
    this.bookings().filter(booking => booking.status === 'confirmed')
  );

  public readonly totalRevenue = computed(() => 
    this.confirmedBookings().reduce((total, booking) => total + booking.totalCost, 0)
  );

  constructor(private http: HttpClient, private authService: AuthService) {
    // Load initial data
    this.loadTurfs();
    this.loadUserBookings();
  }

  /**
   * Get HTTP headers with authorization
   */
  private getAuthHeaders(): HttpHeaders {
    return this.authService.getAuthHeaders();
  }

  /**
   * Load all available turfs from backend
   */
  loadTurfs(sportType?: string, city?: string): Observable<TurfResponse[]> {
    this.isLoading.set(true);
    
    let url = `${this.API_BASE_URL}/booking/turfs`;
    const params: string[] = [];
    
    if (sportType) params.push(`sportType=${sportType}`);
    if (city) params.push(`city=${city}`);
    
    if (params.length > 0) {
      url += '?' + params.join('&');
    }

    return this.http.get<TurfResponse[]>(url)
      .pipe(
        tap(turfs => {
          // Convert backend TurfResponse to frontend Turf
          const convertedTurfs: Turf[] = turfs.map(turf => ({
            id: turf.id.toString(),
            name: turf.name,
            location: turf.address,
            pricePerHour: turf.pricePerHour,
            isAvailable: turf.isAvailable,
            imageUrl: turf.imageUrl
          }));
          
          this.turfs.set(convertedTurfs);
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          return throwError(() => error);
        })
      );
  }

  /**
   * Load user's bookings from backend
   */
  loadUserBookings(): Observable<BookingResponse[]> {
    this.isLoading.set(true);
    
    return this.http.get<BookingResponse[]>(`${this.API_BASE_URL}/booking/my-bookings`, {
      headers: this.getAuthHeaders()
    })
      .pipe(
        tap(bookings => {
          // Convert backend BookingResponse to frontend Booking
          const convertedBookings: Booking[] = (bookings ?? []).map(booking => ({
            id: booking.id.toString(),
            playerName: 'Current User', // This should come from user context
            turfId: booking.turfId.toString(),
            turfName: booking.turfName,
            date: booking.bookingDate,
            timeSlot: booking.startTime,
            duration: booking.durationHours,
            teamSize: 1, // Default value
            totalCost: booking.totalAmount,
            status: booking.status as 'pending' | 'confirmed' | 'cancelled',
            createdAt: new Date(booking.createdAt)
          }));
          
          this.bookings.set(convertedBookings);
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          return throwError(() => error);
        })
      );
  }

  /**
   * Create a new booking
   */
  createBooking(bookingData: CreateBookingRequest): Observable<BookingResponse> {
    this.isLoading.set(true);
    
    return this.http.post<BookingResponse>(`${this.API_BASE_URL}/booking`, bookingData, {
      headers: this.getAuthHeaders()
    })
      .pipe(
        tap(response => {
          // Add the new booking to the local state
          const newBooking: Booking = {
            id: response.id.toString(),
            playerName: 'Current User',
            turfId: response.turfId.toString(),
            turfName: response.turfName,
            date: response.bookingDate,
            timeSlot: response.startTime,
            duration: response.durationHours,
            teamSize: 1,
            totalCost: response.totalAmount,
            status: response.status as 'pending' | 'confirmed' | 'cancelled',
            createdAt: new Date(response.createdAt)
          };
          
          this.bookings.update(bookings => [...bookings, newBooking]);
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get booking by ID
   */
  getBookingById(bookingId: string): Observable<BookingResponse> {
    return this.http.get<BookingResponse>(`${this.API_BASE_URL}/booking/${bookingId}`, {
      headers: this.getAuthHeaders()
    })
      .pipe(
        catchError(error => throwError(() => error))
      );
  }

  /**
   * Cancel a booking
   */
  cancelBooking(bookingId: string): Observable<any> {
    this.isLoading.set(true);
    
    return this.http.post(`${this.API_BASE_URL}/booking/${bookingId}/cancel`, {}, {
      headers: this.getAuthHeaders()
    })
      .pipe(
        tap(() => {
          // Update local state
          this.bookings.update(bookings =>
            bookings.map(booking =>
              booking.id === bookingId
                ? { ...booking, status: 'cancelled' as const }
                : booking
            )
          );
          this.isLoading.set(false);
        }),
        catchError(error => {
          this.isLoading.set(false);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get turf by ID
   */
  getTurfById(turfId: string): Turf | undefined {
    return this.turfs().find(turf => turf.id === turfId);
  }

  /**
   * Get bookings for a specific date
   */
  getBookingsForDate(date: string): Booking[] {
    return this.bookings().filter(booking => booking.date === date);
  }

  /**
   * Get bookings for a specific turf
   */
  getBookingsForTurf(turfId: string): Booking[] {
    return this.bookings().filter(booking => booking.turfId === turfId);
  }

  /**
   * Search turfs with filters
   */
  searchTurfs(filters: { sportType?: string; city?: string; priceRange?: string }): Observable<TurfResponse[]> {
    return this.loadTurfs(filters.sportType, filters.city);
  }
} 

File: playnow-backend/Services/BookingService.cs
using Microsoft.EntityFrameworkCore;
using PlayNow.API.Data;
using PlayNow.API.DTOs;
using PlayNow.API.Models;

namespace PlayNow.API.Services
{
    public interface IBookingService
    {
        Task<BookingResponse> CreateBookingAsync(CreateBookingRequest request, string userId);
        Task<BookingResponse> GetBookingAsync(int bookingId, string userId);
        Task<List<BookingResponse>> GetUserBookingsAsync(string userId);
        Task<bool> CancelBookingAsync(int bookingId, string userId);
        Task<List<TurfResponse>> GetAvailableTurfsAsync(string? sportType = null, string? city = null);
    }

    public class BookingService : IBookingService
    {
        private readonly ApplicationDbContext _context;

        public BookingService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<BookingResponse> CreateBookingAsync(CreateBookingRequest request, string userId)
        {
            // Check if turf exists and is available
            var turf = await _context.Turfs
                .FirstOrDefaultAsync(t => t.Id == request.TurfId && t.IsAvailable && t.IsActive);

            if (turf == null)
            {
                throw new InvalidOperationException("Turf not found or not available.");
            }

            // Check for booking conflicts
            var conflictingBooking = await _context.Bookings
                .AnyAsync(b => b.TurfId == request.TurfId &&
                              b.BookingDate == request.BookingDate &&
                              b.Status != "Cancelled" &&
                              ((b.StartTime <= request.StartTime && b.EndTime > request.StartTime) ||
                               (b.StartTime < request.EndTime && b.EndTime >= request.EndTime) ||
                               (b.StartTime >= request.StartTime && b.EndTime <= request.EndTime)));

            if (conflictingBooking)
            {
                throw new InvalidOperationException("Selected time slot is already booked.");
            }

            // Calculate duration and total amount
            var duration = request.EndTime - request.StartTime;
            var durationHours = (int)Math.Ceiling(duration.TotalHours);
            var totalAmount = turf.PricePerHour * durationHours;

            var booking = new Booking
            {
                UserId = userId,
                TurfId = request.TurfId,
                BookingDate = request.BookingDate,
                StartTime = request.StartTime,
                EndTime = request.EndTime,
                DurationHours = durationHours,
                TotalAmount = totalAmount,
                Status = "Pending",
                SpecialRequests = request.SpecialRequests
            };

            _context.Bookings.Add(booking);
            await _context.SaveChangesAsync();

            return await GetBookingAsync(booking.Id, userId);
        }

        public async Task<BookingResponse> GetBookingAsync(int bookingId, string userId)
        {
            var booking = await _context.Bookings
                .Include(b => b.Turf)
                .Include(b => b.Payment)
                .FirstOrDefaultAsync(b => b.Id == bookingId && b.UserId == userId);

            if (booking == null)
            {
                throw new InvalidOperationException("Booking not found.");
            }

            return new BookingResponse
            {
                Id = booking.Id,
                TurfId = booking.TurfId,
                TurfName = booking.Turf.Name,
                BookingDate = booking.BookingDate,
                StartTime = booking.StartTime,
                EndTime = booking.EndTime,
                DurationHours = booking.DurationHours,
                TotalAmount = booking.TotalAmount,
                Status = booking.Status,
                SpecialRequests = booking.SpecialRequests,
                CreatedAt = booking.CreatedAt,
                Payment = booking.Payment != null ? new PaymentResponse
                {
                    Id = booking.Payment.Id,
                    PaymentMethod = booking.Payment.PaymentMethod,
                    TransactionId = booking.Payment.TransactionId,
                    Amount = booking.Payment.Amount,
                    Status = booking.Payment.Status,
                    PaymentDate = booking.Payment.PaymentDate
                } : null
            };
        }

        public async Task<List<BookingResponse>> GetUserBookingsAsync(string userId)
        {
            var bookings = await _context.Bookings
                .Include(b => b.Turf)
                .Include(b => b.Payment)
                .Where(b => b.UserId == userId)
                .OrderByDescending(b => b.CreatedAt)
                .ToListAsync();

            return bookings.Select(b => new BookingResponse
            {
                Id = b.Id,
                TurfId = b.TurfId,
                TurfName = b.Turf.Name,
                BookingDate = b.BookingDate,
                StartTime = b.StartTime,
                EndTime = b.EndTime,
                DurationHours = b.DurationHours,
                TotalAmount = b.TotalAmount,
                Status = b.Status,
                SpecialRequests = b.SpecialRequests,
                CreatedAt = b.CreatedAt,
                Payment = b.Payment != null ? new PaymentResponse
                {
                    Id = b.Payment.Id,
                    PaymentMethod = b.Payment.PaymentMethod,
                    TransactionId = b.Payment.TransactionId,
                    Amount = b.Payment.Amount,
                    Status = b.Payment.Status,
                    PaymentDate = b.Payment.PaymentDate
                } : null
            }).ToList();
        }

        public async Task<bool> CancelBookingAsync(int bookingId, string userId)
        {
            var booking = await _context.Bookings
                .FirstOrDefaultAsync(b => b.Id == bookingId && b.UserId == userId);

            if (booking == null)
            {
                throw new InvalidOperationException("Booking not found.");
            }

            if (booking.Status == "Cancelled")
            {
                throw new InvalidOperationException("Booking is already cancelled.");
            }

            if (booking.Status == "Completed")
            {
                throw new InvalidOperationException("Cannot cancel completed booking.");
            }

            booking.Status = "Cancelled";
            booking.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<List<TurfResponse>> GetAvailableTurfsAsync(string? sportType = null, string? city = null)
        {
            var query = _context.Turfs
                .Include(t => t.TurfAmenities)
                .Where(t => t.IsAvailable && t.IsActive);

            if (!string.IsNullOrEmpty(sportType))
            {
                query = query.Where(t => t.SportType == sportType);
            }

            if (!string.IsNullOrEmpty(city))
            {
                query = query.Where(t => t.City == city);
            }

            var turfs = await query.ToListAsync();

            return turfs.Select(t => new TurfResponse
            {
                Id = t.Id,
                Name = t.Name,
                Description = t.Description,
                Address = t.Address,
                City = t.City,
                SportType = t.SportType,
                Capacity = t.Capacity,
                PricePerHour = t.PricePerHour,
                ImageUrl = t.ImageUrl,
                IsAvailable = t.IsAvailable,
                Amenities = t.TurfAmenities.Select(ta => ta.Name).ToList()
            }).ToList();
        }
    }
} 